engines/supernova/supernova.h

```diff
@@ -0,0 +1,208 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifndef SUPERNOVA_H
+#define SUPERNOVA_H
+
+#include "audio/audiostream.h"
+#include "audio/mixer.h"
+#include "audio/decoders/raw.h"
+#include "common/array.h"
+#include "common/events.h"
+#include "common/random.h"
+#include "common/scummsys.h"
+#include "engines/engine.h"
+#include "common/file.h"
+#include "common/memstream.h"
+
+#include "supernova/console.h"
+#include "supernova/graphics.h"
+#include "supernova/msn_def.h"
+#include "supernova/rooms.h"
+
+
+namespace Supernova {
+
+#define SAVEGAME_HEADER MKTAG('M','S','N','1')
+#define SAVEGAME_VERSION 8
+
+#define SUPERNOVA_DAT "supernova.dat"
+#define SUPERNOVA_DAT_VERSION 1
+
+
+struct ScreenBuffer {
+	ScreenBuffer()
+		: _x(0)
+		, _y(0)
+		, _width(0)
+		, _height(0)
+		, _pixels(NULL)
+	{}
+
+	byte *_pixels;
+	int _x;
+	int _y;
+	int _width;
+	int _height;
+};
+class ScreenBufferStack {
+public:
+	ScreenBufferStack();
+
+	void push(int x, int y, int width, int height);
+	void restore();
+
+private:
+	ScreenBuffer _buffer[8];
+	ScreenBuffer *_last;
+};
+
+class SupernovaEngine : public Engine {
+public:
+	explicit SupernovaEngine(OSystem *syst);
+	~SupernovaEngine();
+
+	virtual Common::Error run();
+
+	Common::RandomSource *_rnd;
+	GameManager *_gm;
+	Console *_console;
+	Audio::SoundHandle _soundHandle;
+	ScreenBufferStack _screenBuffer;
+	byte _mouseNormal[256];
+	byte _mouseWait[256];
+	MSNImageDecoder *_currentImage;
+	struct SoundSample {
+		byte *_buffer;
```

Comment:
These pointers should be initialized to `nullptr` at `SupernovaEngine` construction time to avoid freeing garbage pointers in the destructor.

Reply:
With regards to memory management, I would even recommend that, throughout, (1) things that can be made members of the engine object itself should (e.g. `SupernovaEngine::_rnd`), since this avoids unnecessary small heap allocations, and (2) anything else which doesn’t need to be copyable and which owns heap-allocated memory (like `SoundSample::_buffer`) should use `Common::ScopedPtr` (for objects) or `Common::Array` (for collections of things, including bytes) and not a raw pointer, since this provides much better guarantees against leaks, missing initialisations, double-frees, and frees of garbage memory.